### DISCUSSION TITLE
Understanding custom "Content Fields" and "Content Parts": when to create custom "Content Fields"?

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15417] [Q&A] [answered]
Hello,

I'm a newbie of Orchard Core and I'm currently diving deeper into Orchard Core to understand best practices. I've been following the "Dojo Course 3" video series, and videos [28](https://youtu.be/bayT58i7DVY) and [29](https://youtu.be/zp1veioH1CQ) clearly explain how to create custom "Content Parts", which is useful for creating reusable components, such as a carousel module for example. I also read the documentation about [custom Content Fields](https://docs.orchardcore.net/en/main/docs/reference/modules/ContentFields/#creating-custom-fields).

For example, as exercise I'm trying to develop a bookshop manager and I need to create a custom "Content Part" for a `Book`. I'd like to implement [custom validation attributes](https://learn.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-8.0#custom-attributes) on its properties, **that utilize a regex**. Additionally, **I want this regex to be customizable from the Admin panel** and perform the validation when the Admin add a new book to the DB. 

**Is this scenario appropriate for a custom "Content Field"?** 

In the past I used Umbraco, where it was very helpful to specify a [validation regex](https://docs.umbraco.com/umbraco-forms/developer/extending/adding-a-validation-pattern#using-the-pattern) for every field of an object I created.

**My question is**: how do you determine when you need a custom "Content Field"? In the "Content Parts", you can define your custom model with your custom types and complexity. So, when is it recommended to break it down into custom smaller pieces, and create custom Content Fields?

Thank you

**Comment by douwinga** [comment, marked as answer]
I am not much of a contributor to OrchardCore, so take this with a grain of salt (but have been using Orchard and OrchardCore for many years). I will explain how I think of things though and maybe it will help.

First, for your regex example, I would just use a [custom editor](https://docs.orchardcore.net/en/latest/docs/reference/modules/ContentFields/#creating-custom-editors) for the existing text field. So then you would have an option like this on text fields
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/95879822-53fd-44ad-9d67-ce191370ad45)

Content Fields are in a way similar to properties on a class, and then parts are the class (only true for parts that are just collections of fields). As a matter of fact you can create a content part class in code that does just that.
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/b07b0298-f28d-45a8-a77c-39624cc0d108)

Basically I think of content fields as a way to display and edit a single value (like a form field, but displays its value on the frontend). The field may be storing more than just the value in the database, but it is still typically to support just displaying and editing a single value for the user.

Content Parts can be as simple as a collection of content fields, or can provide content items of the content type the part is on with additional functionality. For instance, we have a custom part called PrintPart. It provides the functionality for our content types to have a print button on them and then generate a printable PDF from a liquid template. In that case the part isn't even storing any values, and is just providing functionality.

Maybe some examples for custom fields and parts we have created would help?
We don't have many custom fields as the built in OrchardCore fields cover most things, but here are some examples:
Signature field - Allows users to draw their signature and stores it (we use OrchardCore for an app rather than a website, so users can actually edit content items on the frontend where this type of field makes more sense)
Range Field - A slider with a minimum and maximum value
Encrypted Text Field - Just a text field, but the value is stored in the database using the data protection provider, and an audit trail event is added whenever a user unprotects the value to view it

Some of our parts:
AddressPart - Just a collection of content fields to handle a full address (street, state, city, country, zip, etc)
HelpPart - Shows a button on page that provides a way to lookup help information based on the context of the content item
DuplicateDetectionPart - When a content item is saved, a check is done based on configured parameters to see if there is already a matching content item and prevents saving and links to the duplicate instead.
ChangeHistoryPart - Allows to display specified Audit Trail event types in a formatted way on the frontend

Does this help at all?

**Reply by MarGraz** [reply]
@douwinga, thank you for your reply and the time you spent explaining in depth. I really appreciate it 😀
What you've explained is clear, and of course, examples are always useful.

**My understanding of the topic is as follows:**

- **Custom Content Fields**: this is a way to define my own custom type within Orchard Core (a `T` type that is not already present in the "Content Fields"). This custom type can also have multiple properties and can be reused across multiple "Content Parts". The most useful aspect is that you can create your own custom content fields "Modules" (a class library project), which can be installed as "Features" in Orchard Core (either via NuGet or with a direct reference in the same solution as the Web App).

**For example**, a custom "[ISBN](https://en.wikipedia.org/wiki/ISBN)" field, which also includes a checkbox that the Admin can use to show or hide the field on the frontend, looks like this:

```csharp
    public class IsbnFieldViewModel
    {
        public bool IsVisible { get; set; }
        
        public string ValidationRegex { get; set; }
        
        [IsbnFieldValidationAttribute(nameof(ValidationRegex), ErrorMessage = "ISBN does not meet the required regex pattern.")]
        public string Isbn { get; set; }
    }
```
This ISBN type (custom content field) can be used in a `Book` custom "Content Part". I've also added a custom validation attribute `IsbnFieldValidationAttribute` on top, but the "[custom editor](https://docs.orchardcore.net/en/latest/docs/reference/modules/ContentFields/#creating-custom-editors)" approach you mentioned earlier might be better. In my custom validation attribute, I use reflection to get the "ValidationRegex" property value, which I'm not a fan of. The "ValidationRegex" is set from the Orchard Core Admin (`IsbnField.Edit.cshtml`), and users with different roles can't see that field in the backoffice.

- **Custom Content Parts**: this is an object that you can compose and reuse with several content fields (custom or the one provided by Orchard Core). It is also possible to create "Modules" (a class library project) and install them as "Features". The "Content Parts" are objects that can be part of a "Content Type" and displayed on a page or menu.

In conclusion:

1. Is my understanding correct? Or have I completely misunderstood the meanings of "custom content fields" and "custom content parts"?
2. In the [custom editor](https://docs.orchardcore.net/en/latest/docs/reference/modules/ContentFields/#creating-custom-editors), is there a way to show the Regex field in the backoffice only to users with the "Administration" role?

Thank you

**Reply by douwinga** [reply]
I wish I had more Umbraco experience so I can relate things to that, but I only used it briefly many years ago when evaluating Orchard vs Umbraco. 

If I am understanding correctly, I think your definition of content fields is the way I understand it. It is essentially just a field that shows up on the content item when you edit it in the admin (backend / backoffice) and then just displays its value when the content item is viewed on the frontend. You will likely rarely create custom fields as the built in ones cover most use cases. The way I kind of think of it, is a text field is a string, a numeric field is a decimal, a boolean field is a bool, but they are a bit more powerful because they come with an editor and display for me so I don't have to bother with that. At my company we try to use fields as much as possible rather than Custom content parts with standard data types (string, bool, decimal, etc) because it saves a bit of time since once I add that field to the part it is doing the views, saving, loading, etc for us in a standard repeatable way.

Your custom content parts definition seems like I confused you a bit. Custom parts can be more than just a collection of fields. They can provide any functionality you need that content type to have. They can also be reusable and applied to any number of content types. They can also be reusable on the same content type with named parts (Bag part as an example).

Modules are essentially class libraries to contain your custom functionality. Modules can have multiple features in them that you enable/disable in the admin. For example we have a module called `MyCompany.Fields` which contains all of our custom fields, but it has features defined in the Manifest.cs for the different field types. That way different tenants can have access to the different custom fields only if they need them as we run many different tenants with the same code base and each has vastly different needs.

For your ISBN example, I would still not use a custom field for that. I would still use a custom editor for it because in the end it is just storing a string. I also wouldn't create an ISBN custom editor because I like to keep things more generic and reusable than that, so I would stick with just a Regex custom editor. However, do what you think is best for your team as we have a team dedicated to configuring tenants and as I said our tenants are vastly different from each other. We like to provide tools to our configuration team that can work for more than just their current use case so that they don't need code changes as much. Editors can have their own settings. For a regex custom editor, in the settings, you would just have the expression pattern that the value must meet and in the Update of the driver (which would inherit from `ContentFieldDisplayDriver<TextField>` in this case) you see if the input value matches the pattern. If not, add a model state error.

As far as the IsVisible functionality. I wouldn't have that on a specific field type, but make it available to all field types by making a driver that inherits from `ContentPartFieldDefinitionDisplayDriver` called something like `FieldVisibilitySettingsDriver` that handles saving `FieldVisabilitySettings`. And then in an event handler that inherits from `IShapeDisplayEvents`, in the `DisplayingAsync` event, I would check those settings and determine if the field should display or not. We do something like that, and the field is visible or invisible based on the result of a javascript function that is configured on the field settings.

Do note, that you also have access to the [templates ](https://docs.orchardcore.net/en/latest/docs/reference/modules/Templates/) under "Design" in the admin menu. These can override how your shapes/content types will be displayed using Liquid templates. So you could just use templates to build exactly what you want to display, rather than relying on an IsVisible property.

You also have access to [placements](https://docs.orchardcore.net/en/latest/docs/reference/modules/Placements/), which may or may not work for your use case. But they can be used to hide fields.

For displaying a custom editor to only admins you could just add an authorization check to the Edit of your `ContentFieldDisplayDriver` that checks for the SiteOwner permission. If not authorized just return null, otherwise initialize the shape as normal.

**Reply by MarGraz** [reply]
Thank you again for the clear explanation 😀
**Regarding Umbraco**, in [this video](https://www.youtube.com/watch?v=LijuJJvuZV4&t=3m) at minute 3, you can see the setup for validation. It's straightforward, similar to Orchard Core, with "Fields" that support features like autocomplete, mandatory fields, validation, etc.
Yes, of course, "Content Parts" are indeed useful for adding functionalities, as you described. I forgot to mention that.

**Regarding the "Modules" and "Manifest.cs"**, and their different features based on the tenants, are you referring to [registering a Manifest for each resource](https://docs.orchardcore.net/en/main/docs/reference/modules/Resources/#registering-a-resource-manifest)? If not, could you please provide a link to the documentation/video?

**About the** `ContentPartFieldDefinitionDisplayDriver`, it's interesting. I noticed that in the Orchard Core repository, several classes inherit from `ContentPartFieldDefinitionDisplayDriver`, for example [this one](https://github.com/OrchardCMS/OrchardCore/blob/c45aadcb7914efc1d694d79141278f41805dde21/src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/TextFieldHeaderDisplaySettingsDriver.cs#L10). Can you provide a simple example that includes the event handler? Even a simplified one would be great 😊

Thank you

**Reply by douwinga** [reply]
The module manifest is not related to the resource manifest. It looks like it might lack documentation currently. There is mention of it in the code [generation template documentation](https://docs.orchardcore.net/en/latest/docs/getting-started/templates/). 

If you look at the modules in the OrchardCore source you will see some examples.
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/6a173f47-f796-40d8-9062-c92cd3e6844f)

An example of a manifest with multiple features
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/86df0736-de06-463e-a4a8-9413c05753e4)

You will find display drivers for fields and parts. They are what "drive" the generation of the [shapes](https://docs.orchardcore.net/en/latest/docs/reference/core/Placement/#what-is-a-shape) (note: that video is for Orchard 1 which is different, but the concept is the same) that get rendered on the edit and detail pages for the field / part.

In the "what is a shape" video I linked to above, Sebastian jokes about not needing documentation because you have source. That was back in the Orchard 1 days. Documentation was even more lacking back then, it is getting better with OrchardCore, but I still personally find it easier just to look at the source as more documentation is still needed it appears. Basically just find something in the source that is doing something similar to what you are trying to do and see how it does it and do it in your own module. That is how I learned.

If I have time tonight I will see if I can provide an example. It could take a bit to write something up.

**Reply by MarGraz** [reply]
@douwinga, thank you again.

Now I understand what you mean about the `Manifest` class and adding "features" to it, I've already seen it. I'm also familiar with "Shapes" and have read [this discussion](https://github.com/OrchardCMS/OrchardCore/issues/4121#issuecomment-539608731), watched the Orchard Core [Dojo video](https://www.youtube.com/watch?v=yDtUyQpormU), and followed this useful [learning path in the code](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Lombiq.TrainingDemo/StartLearningHere.md#training-sections).

> In the "what is a shape" video I linked to above, Sebastian jokes about not needing documentation because you have source. That was back in the Orchard 1 days. Documentation was even more lacking back then, it is getting better with OrchardCore, but I still personally find it easier just to look at the source as more documentation is still needed it appears. Basically just find something in the source that is doing something similar to what you are trying to do and see how it does it and do it in your own module. That is how I learned.

I agree with you and Sebastian on learning from the code and using code snippets as references. However, in a product like Orchard Core, documentation should serve as a **reference for best practices**. For example, what I did [here](https://github.com/OrchardCMS/OrchardCore/discussions/15417#discussioncomment-8678750) may not follow best practices, but it works. There can be several ways to achieve same result, but there's usually **one best practice**. References should help me follow best practices, and for new users, I think documentation is crucial for understanding these practices.

I'm also contributing to improving the documentation. As a newbie, whenever I encounter unclear information, I open a discussion and then I make a pull request. I'm also writing a quick startup guide covering some basic concepts and a fast learning path, currently missing, which I will publish upon completion.

> If I have time tonight, I'll try to provide an example. It might take a bit to write something up.

**Please, only if you have time**. Pseudocode would also be ok. Otherwise, don't worry, I was just curious to understand.

Thank you so much

**Reply by douwinga** [reply]
Here is how I would implement the custom field editor. I can work on the `IShapeDisplayEvents` example next. In the code examples see the comments as that is where I am trying to explain things. Let me know if something isn't clear

This assumes you already have a module generated and the `Manifest.cs` file configured for your needs, but nothing else.

1. Add a project reference to your new module in your .Web startup project so that your module is loaded when the server starts up
2. In the Views folder of your module add a `_ViewImports.cshtml` file so that some of OrchardCore tag handlers and such will work in your views
 ```csharp
@inherits OrchardCore.DisplayManagement.Razor.RazorPage<TModel>
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
@addTagHelper *, OrchardCore.DisplayManagement
@addTagHelper *, OrchardCore.ResourceManagement
@addTagHelper *, OrchardCore.Contents.TagHelpers
 ```
3. Add the `OrchardCore.ContentFields` nuget package to your module project
4. Add a folder called `Settings` to your module project
5. Add a class called `TextFieldRegExEditorSettings.cs` to the `Settings` folder
```csharp
public class TextFieldRegExEditorSettings
{
    public string Expression { get; set; }
}
```
6. Add a folder called `Drivers` to your module project
7. Add a class called `TextFieldRegExDisplayDriver.cs` to the `Drivers` folder. This is used to add our validation when an attempt to save the field is made.
```csharp
public class TextFieldRegExDisplayDriver : ContentFieldDisplayDriver<TextField>
{
    private readonly IStringLocalizer S;

    public TextFieldRegExDisplayDriver(
        IStringLocalizer<TextFieldRegExDisplayDriver> localizer
    )
    {
        S = localizer;
    }

    // We only override the UpdateAsync method as we are just trying to add some validation when the field is saved
    public override async Task<IDisplayResult> UpdateAsync(TextField field, IUpdateModel updater, UpdateFieldEditorContext context)
    {
        // Only use this updater if the editor type is RegEx
        if (context.PartFieldDefinition.Editor() != "RegEx")
        {
            return Edit(field, context);
        }

        // Update the "Text" value of the "field" from the Request data
        if (await updater.TryUpdateModelAsync(field, Prefix, f => f.Text))
        {
            if (string.IsNullOrWhiteSpace(field.Text))
            {
                return Edit(field, context);
            }

            // Get our RegEx settings for this field
            var expressionSettings = context.PartFieldDefinition.GetSettings<TextFieldRegExEditorSettings>();

            if (!string.IsNullOrWhiteSpace(expressionSettings?.Expression))
            {
                var regex = new Regex(expressionSettings.Expression);

                // If the text value does not match the regex pattern. Having a model state error will cancel the database session so nothing is written
                // and the error message will show to the user
                if (!regex.IsMatch(field.Text))
                {
                    updater.ModelState.AddModelError(Prefix, nameof(field.Text), S["The text does not match the required pattern for {0}.", context.PartFieldDefinition.DisplayName()]);
                }
            }
        }

        return Edit(field, context);
    }
}
```
8. Add a folder called `ViewModels` to your module project
9. Add a class called `TextFieldRegExEditorSettingsViewModel.cs` to the `ViewModels` folder. This viewmodel isn't really necessary as it is the same as the settings model, but you get the idea
```csharp
public class TextFieldRegExEditorSettingsViewModel
{
    public string Expression { get; set; }
}
```
10. Add a class called `TextFieldRegExEditorSettingsDriver.cs` to the `Settings` folder. This file is used for handling the expression setting on the editor settings
```csharp
public class TextFieldRegExEditorSettingsDriver : ContentPartFieldDefinitionDisplayDriver<TextField>
{
    public TextFieldRegExEditorSettingsDriver() { }

    // Builds the settings editor view shape
    public override IDisplayResult Edit(ContentPartFieldDefinition partFieldDefinition)
    {
        // Note the string name. That is the name of the view it will use. In OrchardCore "_" is replaced with "." so the view it will look for is "TextFieldRegExEditorSettings.Edit"
        return Initialize<TextFieldRegExEditorSettingsViewModel>("TextFieldRegExEditorSettings_Edit", model =>
        {
            var settings = partFieldDefinition.GetSettings<TextFieldRegExEditorSettings>();
            model.Expression = settings.Expression;
        })
        .Location("Editor");
    }

    // Handles saving the editor settings
    public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinition partFieldDefinition, UpdatePartFieldEditorContext context)
    {
        // Only save if the RegEx editor is selected
        if (partFieldDefinition.Editor() == "RegEx")
        {
            var model = new TextFieldRegExEditorSettingsViewModel();
            var settings = new TextFieldRegExEditorSettings();

            // Update the model values with what is sent in the Request data
            await context.Updater.TryUpdateModelAsync(model, Prefix);

            settings.Expression = model.Expression;

            // update the settings
            context.Builder.WithSettings(settings);
        }

        return Edit(partFieldDefinition);
    }
}
```
11. Add a view called `TextField-RegEx.Option.cshtml` to the `Views` folder. This is the dropdown option to add to the Editor select.
```csharp
@{
    string currentEditor = Model.Editor;
}
<option value="RegEx" selected="@(currentEditor == "RegEx")">@T["Regular Expression"]</option>
```
12. Add a view called `TextFieldRegExEditorSettings.Edit.cshtml` to the `Views` folder. This is our editor settings fields
```csharp
@model YourModuleNameSpace.ViewModels.TextFieldRegExEditorSettingsViewModel

<div class="field-editor field-editor-regex">
    <div class="mb-3 w-sm-50">
        <label asp-for="Expression">@T["Regular Expression"]</label>
        <input type="text" asp-for="Expression" class="form-control" />
        <span class="hint">@T["The regular expression the input text must match."]</span>
    </div>
</div>
```
13. Update your modules `Startup.cs` to include the new drivers
```csharp
public class Startup : StartupBase
{
    public override void ConfigureServices(IServiceCollection services)
    {
        services.AddScoped<IContentPartFieldDefinitionDisplayDriver, TextFieldRegExEditorSettingsDriver>();

        services.AddContentField<TextField>()
            .UseDisplayDriver<TextFieldRegExDisplayDriver>();
    }
}
```
14. Startup the project and enable your module's feature in the admin.
15. Add a text field to a content type and set its editor to Regular Expression and set the expression pattern and try it out
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/3a1930e4-ea9e-4c44-95fb-1126403df972)

![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/1cf7415c-65e3-47ed-8c06-a2d0599f0228)

Your module will now look something like this
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/fae2d5bb-19b8-4cf3-9631-e684db74df71)

One thing you could add to this so that there is actually a custom editor and not just validation is a view called `TextField-RegEx.Edit.cshtml` that uses a pattern attribute on the input element with your expression. It should get automatically picked up by naming convention and used for the RegEx editor on edit views.

**Reply by douwinga** [reply]
For the visibility functionality, this is how I would implement it. This is a very barebones implementation and you would probably need to expand on it to make it more useful. This is using the same module as before. This assumes you already worked through those steps.

1. Add a class called `ContentFieldVisabilitySettings.cs` to the `Settings` folder
```csharp
public class ContentFieldVisabilitySettings
{
    public bool IsHidden { get; set; }
}
```
2. Add a class called `ContentFieldVisabilitySettingsViewModel.cs` to the `ViewModels` folder.
```csharp
public class ContentFieldVisabilitySettingsViewModel
{
    public bool IsHidden { get; set; }
}
```
3. Add a class called `ContentFieldVisabilitySettingsDriver.cs` to the `Settings` folder. This handles building the shape for the setting as well as saving the value to the settings.
```csharp
public class ContentFieldVisabilitySettingsDriver : ContentPartFieldDefinitionDisplayDriver
{
    public override IDisplayResult Edit(ContentPartFieldDefinition partFieldDefinition)
    {
        return Initialize<ContentFieldVisabilitySettingsViewModel>("ContentFieldVisabilitySettings_Edit", model =>
        {
            var settings = partFieldDefinition.GetSettings<ContentFieldVisabilitySettings>();

            model.IsHidden = settings.IsHidden;
        }).Location("Editor");
    }

    public override async Task<IDisplayResult> UpdateAsync(ContentPartFieldDefinition partFieldDefinition, UpdatePartFieldEditorContext context)
    {
        var model = new ContentFieldVisabilitySettingsViewModel();

        await context.Updater.TryUpdateModelAsync(model, Prefix);

        context.Builder.MergeSettings<ContentFieldVisabilitySettings>(settings =>
        {
            settings.IsHidden = model.IsHidden;
        });

        return Edit(partFieldDefinition);
    }
}
```
4, Add a view called `ContentFieldVisabilitySettings.Edit.cshtml` to the `Views` folder. This is just a view with a checkbox on it so the user can set the Is Hidden value.
```csharp
@model YourNameSpace.Fields.ViewModels.ContentFieldVisabilitySettingsViewModel

<div class="mb-3">
    <div class="form-check">
        <input asp-for="IsHidden" type="checkbox" class="form-check-input">
        <label class="form-check-label" asp-for="IsHidden">@T["Is Hidden"]</label>
    </div>
</div>
```
5. Add a view called `HiddenField.Wrapper.cshtml` to the `Views` folder. A wrapper view does what it says, it wraps the child content in the specified html. In this case it will hide our field if used.
```csharp
<div style="display:none">
	@await DisplayAsync(Model.Metadata.ChildContent)
</div>
```
6. Add a folder called `Handlers` to your module project
7. Add a class called `ContentFieldVisabilityShapeDisplayEvents.cs` to the `Handlers` folder. This is a handler that handles the Displaying event for all shapes. We can use this and short circuit until we find a shape we want to make changes to.
```csharp
public class ContentFieldVisabilityShapeDisplayEvents : IShapeDisplayEvents
{
	public Task DisplayingAsync(ShapeDisplayContext context)
	{
		// only handle edit views
		if (context.Shape.Metadata.DisplayType != "Edit")
		{
			return Task.CompletedTask;
		}

		var partFieldDefinition = context.Shape.GetType()?.GetProperty("PartFieldDefinition")?.GetValue(context.Shape) as ContentPartFieldDefinition;

		if (partFieldDefinition == null)
		{
			return Task.CompletedTask;
		}

		var settings = partFieldDefinition.GetSettings<ContentFieldVisabilitySettings>();

		if (settings == null)
		{
			return Task.CompletedTask;
		}

		if (settings.IsHidden)
		{
			// if hidden, add a wrapper that just wraps the field in a hidden div
			context.Shape.Metadata.Wrappers.Add("HiddenField_Wrapper");
		}

		return Task.CompletedTask;
	}

	public Task DisplayedAsync(ShapeDisplayContext context)
	{
		return Task.CompletedTask;
	}

	public Task DisplayingFinalizedAsync(ShapeDisplayContext context)
	{
		return Task.CompletedTask;
	}
}
```
8. Update your modules `Startup.cs` to include the new driver and shape display event handler
```csharp
services.AddScoped<IContentPartFieldDefinitionDisplayDriver, ContentFieldVisabilitySettingsDriver>();
services.AddScoped<IShapeDisplayEvents, ContentFieldVisabilityShapeDisplayEvents>();
```

You will now have a setting on all field types. When it is selected our HiddenField wrapper wraps the field and therefore hides it.
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/81165c80-0710-4121-b9a1-b4f7bca84547)

Our module will now look something like this:
![image](https://github.com/OrchardCMS/OrchardCore/assets/11961383/5a47e18a-1fec-41c6-9a78-f5256b5ebbe9)

I setup a [repo](https://github.com/douwinga/orchardcore-tutorials) that has all the code I used for this.

**Reply by MarGraz** [reply]
@douwinga, thank you for the effort you've put into writing those examples, they are really good. Why not include them in the "[How to guides](https://docs.orchardcore.net/en/latest/docs/guides/)" section? If you're unable to do so, I can create a pull request and submit it. I think both examples are excellent starting points. Please, publish them.

From the [first example](https://github.com/OrchardCMS/OrchardCore/discussions/15417#discussioncomment-8686457), point number 10 is particularly interesting. I've never tried it before, but I will.

The [second example](https://github.com/OrchardCMS/OrchardCore/discussions/15417#discussioncomment-8688326) is also very helpful. I'm curious to see how it applies to other fields, I will try it.

Thank you so much, really! 😊

**Reply by douwinga** [reply]
I don't think these would be good enough to go into an official how to guide. The first example should probably use a `ContentFieldHandler` and in the `ValadatingAsync` event it would do the validation instead of using the `TextFieldRegExDisplayDriver.cs` driver just for validation. I couldn't get that to work though, but something you should look into if you are going to do something similar. I haven't used the ContentFieldHandler Validating event in any of my own code since ContentFieldHandlers didn't always exist and we actually had our own implementation of them, but it does not include the Validating event.

Also, the first example should implement the actual editor view as I mentioned with the `TextField-RegEx.Edit.cshtml` so it truly is a custom editor. I am just using an older version of OrchardCore and there have been some changes that would affect how I would do that view if I was on the latest and I didn't want things to be too confusing.

In the second example I do not like `var partFieldDefinition = context.Shape.GetType()?.GetProperty("PartFieldDefinition")?.GetValue(context.Shape) as ContentPartFieldDefinition;` but I think it is the safest way to get the value. You could also cast it to a dynamic object `((dynamic)context.Shape).PartFieldDefinition` I believe and get the value that way, but I think that could break. And since this executes on every shape you have to make it as safe as possible. Maybe a different example that does not need something like that would make more sense in official documentation.

Do remember, these are how I would implement things, I am not saying these are the best ways, or the way OrchardCore maintainers would do things. And as you said, the documentation should have some best practices and show the "correct" way of doing things to guide people.

**Reply by MarGraz** [reply]
**We can open two other discussions and ask the community for contributions on these two examples before pulling them**. What do you think? I believe it's a great way to enrich the documentation 😊
I will certainly keep both examples as a good starting point.

**Regarding the first example**, about `TextFieldRegExDisplayDriver.cs`, if you're considering moving the validation part out of the "display driver" class to adhere to the "single responsibility principle", I think you could adopt a similar approach to what I have done [here](https://github.com/OrchardCMS/OrchardCore/discussions/15417#discussioncomment-8678750), by using a "custom attribute" in `IsbnFieldViewModel`. The main challenge in your first example is retrieving the user input value and applying the custom attribute to that property. **Perhaps we could extend the Orchard Core `TextField` class?**

**This is a concept, I haven't tried running it yet.** 
Here's a concept for the custom validation attribute `ValidateWithRegExSettingAttribute` class, to be created in a new "Helpers" > "DataAnnotations" folder within your module project.

```csharp
[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field, AllowMultiple = false)]
public class ValidateWithRegExSettingAttribute : ValidationAttribute
{
    protected override ValidationResult IsValid(object value, ValidationContext validationContext)
    {
        // Here, you would need to access your settings to get the regex expression.
        // This is purely conceptual and might not work as expected in Orchard Core
        var settings = (TextFieldRegExEditorSettings)validationContext.GetService(typeof(TextFieldRegExEditorSettings));
        var expression = settings.Expression;
        var input = value as string;

        if (!string.IsNullOrWhiteSpace(expression) && !string.IsNullOrWhiteSpace(input))
        {
            var regex = new Regex(expression);
            if (!regex.IsMatch(input))
            {
                return new ValidationResult(ErrorMessage ?? "The input does not match the required pattern.");
            }
        }

        return ValidationResult.Success;
    }
}
```
Here how to apply the custom attribute to the input property. **The challenge is to apply this attribute to a property that holds the user input to be validated**. Since `TextFieldRegExEditorSettings` doesn't hold user input, this attribute would be applied elsewhere, in a context where the user input is available and the settings can be accessed. But where? 🤔

```csharp
public class TextModel
{
    [ValidateWithRegExSetting(ErrorMessage = "The input does not match the required pattern.")]
    public string MyTextField { get; set; }
}
```

What do you think?

**Reply by douwinga** [reply]
To be honest, I don't have the motivation to help create an official guide as it just doesn't interest me enough. I would rather write about things that interest me, but wouldn't belong in the official docs like how to integrate Hangfire into OrchardCore, or how our dynamic index table feature works that allows users to "index" anything on content items to be used in queries. I hope to get to that some day on my blog.

I also am hoping to take contributing code to OrchardCore a bit more serious. So much to do, so little time, but I would like to give back where I am able to since I have been working with Orchard daily for years as the base for our apps. If only I could get work to pay for a bit of my time each week to give back 😜.

As for your attribute idea, I think you have already discovered the issue with that approach. It would be difficult to make extendable enough to work with all of the different editors.

**Reply by MarGraz** [reply]
Yes, it make sense. I love to participate to open source projects, and give back what I took. I'm new of Orchard Core, and I think it has really good potentiality, but it lacks of updated and best practice documentation. Ok that I can learn on the code, but today when you need to use something to create a project, you don't have one month to first learn all the code base, you need a best practice reference.

> As for your attribute idea, I think you have already discovered the issue with that approach. It would be difficult to make extendable enough to work with all of the different editors.

If I find a nice way to do it, I will reply in this discussion.

Thank you so much 😊
